semantic-python/src/Language/Python/Core.hs

```diff
@@ -32,21 +35,44 @@ type CoreSyntax sig t = ( Member Core sig
 
 class Compile py where
   -- FIXME: we should really try not to fail
-  compile :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name)
+  compile :: ( CoreSyntax syn t
+             , Member (Reader Path.RelFile) sig
+             , Carrier sig m
+             , MonadFail m
+             )
+          => py
+          -> m (t Name)
 
   default compile :: (MonadFail m, Show py) => py -> m (t Name)
   compile = defaultCompile
 
-  compileCC :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name) -> m (t Name)
-
-  default compileCC :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name) -> m (t Name)
+  compileCC :: ( CoreSyntax syn t
+               , Member (Reader Path.RelFile) sig
+               , Carrier sig m
+               , MonadFail m
+               )
+            => py
+            -> m (t Name)
+            -> m (t Name)
+
+  default compileCC :: ( CoreSyntax syn t
+                       , Member (Reader Path.RelFile) sig
+                       , Carrier sig m
+                       , MonadFail m
+                       ) => py -> m (t Name) -> m (t Name)
   compileCC py cc = (>>>) <$> compile py <*> cc
 
-locate :: (HasField "ann" syntax Span, CoreSyntax sig t) => syntax -> t a -> t a
-locate syn = Core.annAt (locFromTSSpan (getField @"ann" syn))
-  where
-    locFromTSSpan (TreeSitter.Span (TreeSitter.Pos a b) (TreeSitter.Pos c d))
-      = Data.Loc.Loc mempty (Data.Loc.Span (Data.Loc.Pos a b) (Data.Loc.Pos c d))
+locate :: ( HasField "ann" syntax Span
+          , CoreSyntax syn t
+          , Member (Reader Path.RelFile) sig
+          , Carrier sig m
+          ) => syntax -> t a -> m (t a)
+locate syn item = do
+  fp <- asks @Path.RelFile (Text.pack . Path.toString)
+  let locFromTSSpan (TreeSitter.Span (TreeSitter.Pos a b) (TreeSitter.Pos c d))
+        = Data.Loc.Loc fp (Data.Loc.Span (Data.Loc.Pos a b) (Data.Loc.Pos c d))
```

Comment:
My eventual goal is for `semantic-core` to get its `Span` type from a package which both it and `tree-sitter` can depend on, and thus avoid this translation _and_ avoid having to make `semantic-core` depend on `tree-sitter`.

That’s a nice dream… right? 😆

In the meantime, 👍 to this approach.

Reply:
Re the `Reader Text` versus `Reader RelFile`: how about a newtype wrapper?

```
newtype SourcePath = SourcePath Text deriving IsString
```

Then at the OS-concerned RelFile -> SourcePath boundary (which we will only cross once) we can do `fromString @SourcePath . Path.toString`, which seems readable to me.

Reply:
Love it!